frb/analysis/kcwi.py

```diff
@@ -0,0 +1,362 @@
+
+    sky = np.zeros_like(speclist)
+
+    # Convert wavelengths from air to vacuum
+    wave = cube.spectral_axis.value
+    if tovac:
+        wave = _air_to_vac(wave)
+    wavelist = np.tile(wave, (nobjs,1))
+
+    # Set infs to nan
+    speclist[np.isinf(speclist)] = np.nan
+    varspeclist[np.isinf(varspeclist)] = np.nan
+
+    # Create HDUs
+    extnames = ['INTENSITY', 'VARIANCE', 'SKY', 'WAVELENGTH']
+    datalists = [speclist, varspeclist, sky, wavelist]
+
+    marz_hdu = fits.HDUList()
+    for ext, data in zip(extnames, datalists):
+        hdu = fits.ImageHDU(data)
+        hdu.header.set('extname', ext)
+        marz_hdu.append(hdu)
+
+    # Create object table
+
+    ids = np.arange(nobjs)+1
+    x = objects['x'].data
+    y = objects['y'].data
+    coords = wcsutils.pixel_to_skycoord(x,y,wcsinfo)
+    ra = coords.ra.value
+    dec = coords.dec.value
+    types = ('P_'*nobjs).split('_')[:-1]
+
+    cols = []
+    colnames = ['source_id', 'RA', 'DEC', 'X', 'Y', 'TYPE']
+    formats = ['80A', 'D', 'D', 'J', 'J', '1A']
+    coldata = [ids, ra, dec, x, y, types]
+    for cname, form, cdat in zip(colnames, formats, coldata):
+        cols.append(fits.Column(cname, form, array = cdat))
+    coldefs = fits.ColDefs(cols)
+    tabhdu = fits.BinTableHDU.from_columns(coldefs)
+    tabhdu.header.set('extname', 'FIBRES')
+    marz_hdu.append(tabhdu)
+
+    marz_hdu.writeto(marzfile, overwrite = True)
+    return
+
+def get_source_spectra(cubefile, varfile, objects, outdir = "spectra/", marzfile = None, tovac = True):
+    """
+    Extract spectra of sources found using SExtractor
+    from datacube.
+    Args:
+        cubefile (str): A datacube fits file
+        varfile (str): Variance datacube fits file
+        objects (Table): Table of extracted objects produced
+            by sep.extract
+        outdir (str, optional): directory to store spectra
+        marzfile (str, optional): name of MARZ file to dump
+            all spectra into. File creation is skipped if
+            a name is not supplied.
+    Returns:
+        speclist (ndarray): A 2D array of with an extracted
+            spectrum in each row.
+        varspeclist (ndarray): Similarly designed array with
+            variance information.
+        wave (1D Quantity array): Wavelength array.
+    """
+
+    # Preliminaries
+    nobjs = len(objects)
+    cube = SpectralCube.read(cubefile)
+    varcube = SpectralCube.read(varfile)
+
+    wave = cube.spectral_axis.value
+
+    # Convert to vacuum wavelengths?
+    if tovac:
+        wave = _air_to_vac(wave)
+    # Prepare an HDU in advance
+    wavehdu = fits.ImageHDU(wave)
+    wavehdu.header.set('extname', 'WAVELENGTH')
+
+    # Initialize output lists 
+    speclist = np.zeros([nobjs, len(wave)])
+    varspeclist = np.zeros_like(speclist)
+
+    # Create output folder?
+    if not os.path.isdir(outdir):
+        os.mkdir(outdir)
+    for idx, obj in enumerate(objects):
+        # TODO: Make a specdb file instead of saving
```

Comment:
Individual spectra are fine.  We can make the `specdb` file after.

Reply:
Sounds good!